### ISSUE TITLE
Adding support for dispatching Sagas from Action Creators (like in redux-thunk)

### ISSUE THREAD

**Issue by wizardzloy** [yelouafi/redux-saga#123]
Is there a functionality we lose if we would simply do `dispatch(mySaga)` instead of registering all the sagas in middleware on the store creation step

**Comment by aikoven** [comment]
Something like
```js
function *saga() {
  while (true) {
    yield take([action1, action2])  // watch multiple actions
    // ...
  }
}
```

Or even

```js
    yield take(action => /* ... */)
```

**Comment by ALF-er** [comment]
Also, who will call `mySaga.next()` after each side-effect (`put`, `call` etc)?

**Comment by wizardzloy** [comment]
@aikoven
- do you really need to use `take` if you're able to trigger saga execution by dispatching it? 
- if so, I still don't see why it's not possible to have that effect working with dispatched saga. Same mechanism of subscribtion to internal action emmiter would perfectely work,  wouldn't it? 

@ALF-er 
side effects would be executed by the same middleware as now. I am not talking about getting rid of middleware

**Comment by aikoven** [comment]
@wizardzloy You are right, this would work and IMO it's a valid use case. Still daemon saga is more flexible approach, e.g. you can do things like error handling:

```js
function *errorHandlerSaga() {
  while (true) {
    const action = yield take(action => action.error)
  }
}
```

If you do this via action creators, you would have to trigger the saga manually from every action creator for error action.

**Comment by yelouafi** [comment]
@wizardzloy Dispatching Sagas from Components will *scope* the Saga by the triggering action. Sometimes Sagas need a much higher level view for control flow for example

```javascript
function* saga() {
  while(true) {
     yield take(LOGIN)
     ...authorize logic
     yield take(LOGOUT)
  }
}
```

Using the dispatch on each action method, you'll have to write 2 Sagas for each action and then coordinate the 2 using some stored state in the Store. 

The daemon method also decouples the Component from the Saga, your Action Creators are just pure functions which return plain Objects (and maybe you don't need them anymore). The Component just communicates  what happens.

**Comment by yelouafi** [comment]
I'd like to add the point of `take` is transform the push style of events into a pull style. This way event handling can be embedded inside a synchronous like algrorithm inside Sagas. Somewhat similar to when you use `var char = readLine()` in apps that read from the standard input.

**Comment by wizardzloy** [comment]
@yelouafi 
I totally see your point and agree that for certain situations the "daemon" approach fits better. But I was working on the redux-based application for quite a while and from my personal experience those scenarious you describe are somewhere around 10% of all use cases. They are important, but I really worry about the fact that all my business logic is detached from the views that call for it.
I have an impression that we are moving towards eventEmitter approach and I am not a huge fan of that.

So basically the question I really wanted to ask is would the PR that adds support of dispatching sagas be appreciated? Just for sake of giving people an option to choose either push or pull approach. Or maybe the daemons approach is the only one that library wants to enforce,  which is perfectly fine as well btw)

**Comment by yelouafi** [comment]
>They are important, but I really worry about the fact that all my business logic is detached from the views that call for it.

Can you elaborate on this ? At a first glance it seems like a good thing to have this kind of decoupling. But I'd like to hear you opinion

>Or maybe the daemons approach is the only one that library wants to enforce

From API POV I'm not a fan of enforcing things. I don't like to dictate to people how they write their apps.

My question, what are there pros/cons from introducing the support of dispatching Sagas ? 

Would love to hear what other people think; (cc @slorber, @aikoven, @gaearon)

One benefit I'm  seeing actually is that it would make transitioning from redux-thunk easier: Generators in this case will offer a more testable alternative to async functions.

OTOH I'm wondering if this would make people miss the point of Sagas, because the dispatch approach is easier and would likely be used widely (But As I said that's an opinion and I don't like to dictate it on people)

**Comment by ntkoso** [comment]
I think main point of redux-saga is total isolation of components from side effects. With ability to dispatch sagas you effectively enable ability to dispatch side effects. And people will try to use redux-saga the same way they are using redux-thunk, redux-promise, etc... 

Some confusion for new users in this case:

```javascript
// some ./src/actions.js
export function fetchSomething(...args) {
  return function* () {
    return yield call(fetch, ...args);
  }
}

export function clickFetch() {
  return function* () {
    const result = yield put(fetchSomething(...args)) // put === dispatch ?
    // or
    const result = yield call(fetchSomething(...args)) // put vs call ? put vs fork?
  }
}
```

They will miss whole point that sagas operate on single 'reduxStoreActions' channel. And then will ask why `put` named 'put' instead of 'dispatch'.

Idea of flux was in clear separation of business logic and views using action log. 
And `redux-saga` takes it to the new level with 'daemons', making action creators pure.

**Comment by wizardzloy** [comment]
@yelouafi
> Can you elaborate on this ? At a first glance it seems like a good thing to have this kind of decoupling. But I'd like to hear you opinion

It sounds good theoretically, but it's not that easy practically. We actually have all same problems as with eventEmitter based architecture. It becomes hard or (for really large apps) even impossible to find all the places that react to certain event. We dont have single entrypoint for your business logic anymore, instead application responds to your event from different places (multiple different sagas) which makes it hard to reason about the scope of that response. 
You may say that we could create sort of one-to-one relationship between Action and Saga to be sure that we could easily find the corresponding Saga and don't bother searching through the entire app. But in this case we effectively getting back to the dispatch(saga) approach but with artificially created indirection (matching Action with our Saga)

**Comment by slorber** [comment]
> OTOH I'm wondering if this would make people miss the point of Sagas, because the dispatch approach is easier and would likely be used widely (But As I said that's an opinion and I don't like to dictate it on people)

I totally agree. The idea of sagas for me is to orchestrate operations across decoupled components. Triggering sagas in an imperative way seems weird to me, and is not even close to backend saga concept at all. 

The problem is that we are mixing kind of 2 different things in this lib. Sagas is a pattern from backend world. This library permits to implement the saga pattern for frontend world, but not only that. So redux-saga should rather choose if it wants to be a more general toolkit for side-effects, permitting to implement the saga pattern, or if it wants to enforce a clean usage of the saga pattern.

Because for me, dispatching sagas, or running local sagas on component mount is not the saga pattern anymore and I'm not planning to use such things at all.

**Comment by wizardzloy** [comment]
@slorber fair enough. Maybe it makes sense to create a different library for those who prefer `redux-thunk` approach but with all goodies of declarative side effects and `call`, `fork/join`, `put` operators. Because trying to satisfy everyone we might break the library's consistency. 

One question though, @slorber, was my analogy with EventEmitter right or maybe I just didn't get the idea?

**Comment by slorber** [comment]
I'm not sure to understand that analogy sorry.

The saga pattern is the coupling point between different decoupled bounded contexts (you can compare a bounded context to a widget). Widgets should not know about each others but just know about themselves. The saga is the piece that says "when button is clicked inside widget1, then widget2 should toggle", without widget1 having to know anything about widget2. 
You don't create a 1-1 mapping between an widget or an action to a saga. The widget should not even be aware of any saga and should be able to work fine without that saga (only the global cooperation of widgets should be affected).

Yes this creates indirection that is overkill in small projects but will make the app scale better on big projects.

You may or may not use redux-saga this way, but this is what the saga pattern is initially.

**Comment by yelouafi** [comment]
@wizardzloy I think you can achieve what you want with the actual API using the dynamic running feature

`configureStore.js`
```javascript
import createSagaMiddleware from 'redux-saga'
import { createStore, applyMiddleware } from 'redux'
import reducer from '...'

const sagaMiddleware = createSagaMiddleware()
export const runSaga = sagaMiddleware.run

export default configureStore() {
  return createStore(
    reducer,
    applyMiddleware(/*..*/, sagaMiddleware)
  )
}
```

`actions.js`
```javascript
import { runSaga } from './configureStore'

export function fetchSomething(url) {
  runSaga(function*() {
     const response = yield call(api.fetch, url)
     yiedl put(success(response))
  })  
}
```

Maybe you can even create a middleware for that

**Comment by wizardzloy** [comment]
@slorber thank you for explanation, it indeed reminds me of Backbone/Marionette apps that were highly relying on EventBus communication between components.

@yelouafi thanks for examples, I'll try that

**Comment by paul42** [comment]
I'm a complete novice, but coming from backend programming I just want to say I like the saga model, and I would prefer it to thunk, even sacrificing simplicity of setup because I like the api around sagas. I feel that @wizardzloy does have a valid point that we can't see into and around and know who is calling what but I'm wondering if that's more of a tooling problem rather than anything - there was a very interesting .gif I saw from the cycle.js team (guy?) where he dispatched an action and it had a moving diagram I have zero clue how it was implemented/drawn but I think adding a visual component might help some people - rambling aside I wanted to underscore that I like the saga model and I think tooling can fix the proposed issue without modifying the base library. (redux-devtools-visual-map?)

**Comment by aikoven** [comment]
I agree that it may be hard to trace when and where an action was dispatched when your app grows. 
Still I prefer to isolate every possible side-effect in Sagas and keep action creators pure.

To address this specific issue I'm using following trick:
I have action creator factory that looks something like this:
```js
function actionCreator(type) {
  return payload => ({type, payload})
}
```
Because all my action creators are pure, I'm using this factory for all of them. Now I can use `Error` to add stack trace to every action:
```js
function actionCreator(type) {
  return payload => {
    let stack;
    try {
      throw new Error()
    } catch (e) {
      stack = e.stack;
    }
    return {
      type, payload,
      meta: {stack}
    }
  }
}
```
With little help from [error-stack-parser](https://github.com/stacktracejs/error-stack-parser) I can get the following result:

![screen shot 2016-02-19 at 21 15 58](https://cloud.githubusercontent.com/assets/857709/13179442/dc7eaf8c-d745-11e5-9328-7737e885e296.png)

**Comment by wizardzloy** [comment]
@aikoven nice trick! 

Regarding:
> Still I prefer to isolate every possible side-effect in Sagas and keep action creators pure.

I don't see how the fact that you could dispatch sagas contradicts with idea of having pure action creators.

**Comment by kloy** [comment]
:+1: 

Dispatching generators like how thunks work is something I found to be missing as well. I understand that the same goal can be accomplished by triggering sagas inside of sagas. This however seems like additional boilerplate with little value for a good portion of use cases. My common use case for generators is to coordinate asynchronous behaviors for submitting forms. In traditional submit -> wait for response while showing some type of spinner forms a saga that is taking a submit action then starting a sub saga for the request seems like over kill. This is a very short lived saga and a more obvious implementation if dispatched as part of an onClick handler.

I agree sagas taking actions is a good approach, and I am in no way advocating for the removal of this behavior. I am advocating for when a generator is dispatched the saga middleware handles and executes it as a new framework feature. By keeping these short lived processes closer to the initialization code (onClick -> dispatch action) you keep the code base simpler and easier to triage as it is declarative. If all sagas must wait for an action you add the additional overhead of first tracking down what action (event) is fired then searching in the codebase for that action type to locate the saga operating on it. If you have multiple sagas listening on the same action types this of course adds even more overhead of search for the effect. This quickly starts to smell like event soup when applied to all effects in a UI code base.

**Comment by ganarajpr** [comment]
@wizardzloy I can understand where you are coming from - and I had the same initial reaction with saga. It feels soo much like emitting an event and catching it somewhere else to do something. BUT.. the real power of saga comes in because it allows you to specify complete user interaction flows as a single function / generator. This is something you cannot do - as elegantly with event emitters is my guess. This is mostly enabled because of the control flow constructs provided by the generators. 

Here is an example 
```
function * onEditAddress(action) {
  yield put(showAddressFormDialog());
  const { success } = yield race({
    success: take(SAVE_ADDRESS_SUCCESS),
    error: take(SAVE_ADDRESS_ERROR),
    close: take(HIDE_ADDRESSFORM_DIALOG)
  });
  if (success) {
    yield put(hideAddressFormDialog());
  }
  // ... further control flow logic..
}

export function * watchStartEditAddress() {
  yield* takeLatest(START_EDIT_ADDRESS, onEditAddress);
}

```

By just seeing this generator its quite clear how the application flow is structured. This same elegance is not possible with event emitters as far as I can tell. Let me know if you think otherwise.

**Comment by wizardzloy** [comment]
@ganarajpr man, I totally enjoy the generator based API for control flow and declarative side effects, but I think @kloy described exactly what are my concerns there. Sometimes having sagas as daemons is beneficial (e.g. for long lived business logic flow, like maybe login/logout flow). But most of the time on the front end you use short lived sagas that are spawned, executed and killed within couple of seconds. For latter I believe dispatching sagas is a better alternative, as an additional lib feature of course

**Comment by yelouafi** [comment]
>But most of the time on the front end you use short lived sagas that are spawned, executed and killed within couple of seconds. 

I don't think it's that simple. The 'easy' benefit of the dispatch approach tend to hide some subtle issues from the developer.

 If we take what seems to be the most common use case which is ajax requests. Dispatching async things (thunks, generators, promises) seems to be the easiest approach but it hides subtle concurrency issues (out of order responses, which can be handled by the daemon approach). New users will likely be unaware of those issues.

UI applications are inherently concurrents, and even what seems to be an isolated action is likely to be a part of a much larger flow. The daemon/pull approach makes those requirements more visibles : When you start forking tasks you are more exposed to the issue of coordinating those tasks.

There are some benefits of the dispatch approach, but if this is going to be implemented. I prefer it to be a separate middleware. 

```javascript
const createSagaDispatcher = sagaMiddleware => store => next => action => {
   if(is.iterator(action)) {
     return sagaMiddleware.run(action).done
   }
   return next(action)
}
```

Usage

```javascript
import { createStore, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import createSagaDispatcher from 'redux-saga-dispatcher'

import reducer from './reducer'
import daemonSagas from './saga'

const sagaMiddleware = createSagaMiddleware(...daemonSagas)
const sagaDispatcher  = createSagaDispatcher(sagaMiddleware)

const store = createStore(
  reducer,
  applyMiddleware( sagaDispatcher , sagaMiddleware)
)
```

If you want to make this a separate package, I don't mind linking it from the redux-saga docs.

**Comment by kloy** [comment]
@yelouafi Thank you for the example. Below I have pasted a working middleware.

```javascript
function isGenerator(fn) {
  return typeof fn === 'function' && 'next' in fn.prototype && 'throw' in fn.prototype;
}

const createSagaDispatcher = sagaMiddleware => () => next => action => {
  if (isGenerator(action)) {
    return sagaMiddleware.run(action).done;
  }

  return next(action);
};

export default createSagaDispatcher;
```


Here is an example of dispatching a generator...

https://github.com/kloy/saga-playground/blob/master/src/index.js#L33

This is not meant as a good example of when to dispatch a generator, just a way for me to prove the functionality.

**Comment by yelouafi** [comment]
I'll close this. As I don't plan to add dispatching to the core. Let me know If someone would like to contribute a separate middleware for this.
